### PR TITLE
Flush BufferedOutput on close

### LINKED ISSUE #578: Buffered output not flush()ed before close()
@leandro-lucarella-sociomantic :

This is a pretty nasty bug:

``` d
import ocean.io.Stdout;
import ocean.io.stream.Buffered;
import ocean.io.device.File;

void main()
{
    auto output = new BufferedOutput(new File("test.txt", File.WriteCreate));
    output("hello\n");
    version (Flush) output.flush();
    output.close();
}
```

When compiled without `-version=Flush` it produces an empty file! I don't see anything in the docs referring to this behaviour, and even if it is, is completely wrong and unintuitive!

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Fixes #578

### CODE REVIEW

File: src/ocean/io/stream/Buffered.d [hunk 1]
```diff
@@ -1166,6 +1166,18 @@ public class BufferedOutput : OutputFilter, OutputBuffer
         return this;
     }
 
+    /***************************************************************************
+
+        Flushes the buffer and closes the stream.
+
+    ***************************************************************************/
+
+    final override void close ( )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/stream/Buffered.d:1175]: Would be nice to add a unittest for this.

Comment (mihails-strasuns) [reply src/ocean/io/stream/Buffered.d:1175]: That would be inherently flakey because you don't control buffering afaics.

Comment (gavin-norman-sociomantic) [reply src/ocean/io/stream/Buffered.d:1175]: Hmm, is there some way that the buffer would be flushed magically behind the scenes?

I was thinking of something like:
1. Open file with buffered output.
2. Write something smaller than the buffer size.
3. Close the file.
4. Check that the file content on disk is correct. (i.e. the buffer was flushed.)

Comment (mihails-strasuns) [reply src/ocean/io/stream/Buffered.d:1175]: Yeah, makes sense. Added.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**